File: CPP/hwopt/hwopt/Noise/ImprovedNoise.cpp
#include "ImprovedNoise.h"

ImprovedNoise::ImprovedNoise() {
    std::random_device rd;
    std::mt19937_64 mt(rd());
    std::uniform_real_distribution dist_double(0.0, 1.0);

    this->xo = dist_double(mt) * 256.0;
    this->yo = dist_double(mt) * 256.0;
    this->zo = dist_double(mt) * 256.0;

    for (int i = 0; i < 256; i++) {
        this->p[i] = static_cast<char>(i);
    }

    for (int i = 0; i < 256; i++) {
        std::uniform_int_distribution d(0, 256 - i);
        const int offset = d(mt);
        const char tmp = this->p[i];
        this->p[i] = this->p[i + offset];
        this->p[i + offset] = tmp;
    }
}

auto ImprovedNoise::noise(const double _x, const double _y, const double _z) const -> double {
    return this->noise(_x, _y, _z, 0.0, 0.0);
}

auto ImprovedNoise::noise(const double _x, const double _y, const double _z, const double yScale, const double yFudge) const -> double {
    const double x = _x + this->xo;
    const double y = _y + this->yo;
    const double z = _z + this->zo;
    const int xf = std::floor(x);
    const int yf = std::floor(y);
    const int zf = std::floor(z);
    const double xr = x - xf;
    const double yr = y - yf;
    const double zr = z - zf;
    double yr_fudge;
    if (yScale != 0.0) {
        double fudge_limit;
        if (yFudge >= 0.0 && yFudge < yr) {
            fudge_limit = yFudge;
        } else {
            fudge_limit = yr;
        }

        yr_fudge = std::floor(fudge_limit / yScale + 1.0E-7F) * yScale;
    } else {
        yr_fudge = 0.0;
    }

    return this->sample_and_lerperm(xf, yf, zf, xr, yr - yr_fudge, zr, yr);
}

auto ImprovedNoise::noise_with_derivative(const double _x, const double _y, const double _z, double* derivativeOut) const -> double {
    const double x = _x + this->xo;
    const double y = _y + this->yo;
    const double z = _z + this->zo;
    const int xf = std::floor(x);
    const int yf = std::floor(y);
    const int zf = std::floor(z);
    const double xr = x - xf;
    const double yr = y - yf;
    const double zr = z - zf;
    return this->sample_with_derivative(xf, yf, zf, xr, yr, zr, derivativeOut);
}

auto ImprovedNoise::grad_dot(const int hash, const double x, const double y, const double z) -> double {
    return SimplexNoise::dot(SimplexNoise::GRADIENT[hash & 15], x, y, z);
}

auto ImprovedNoise::perm(const int x) const -> int {
    return this->p[x & 0xFF] & 0xFF;
}

auto ImprovedNoise::sample_and_lerperm(const int x, const int y, const int z, const double xr, const double yr, const double zr, const double yrOriginal) const -> double {
    const int x0 = this->perm(x);
    const int x1 = this->perm(x + 1);
    const int xy00 = this->perm(x0 + y);
    const int xy01 = this->perm(x0 + y + 1);
    const int xy10 = this->perm(x1 + y);
    const int xy11 = this->perm(x1 + y + 1);
    const double d000 = grad_dot(this->perm(xy00 + z), xr, yr, zr);
    const double d100 = grad_dot(this->perm(xy10 + z), xr - 1.0, yr, zr);
    const double d010 = grad_dot(this->perm(xy01 + z), xr, yr - 1.0, zr);
    const double d110 = grad_dot(this->perm(xy11 + z), xr - 1.0, yr - 1.0, zr);
    const double d001 = grad_dot(this->perm(xy00 + z + 1), xr, yr, zr - 1.0);
    const double d101 = grad_dot(this->perm(xy10 + z + 1), xr - 1.0, yr, zr - 1.0);
    const double d011 = grad_dot(this->perm(xy01 + z + 1), xr, yr - 1.0, zr - 1.0);
    const double d111 = grad_dot(this->perm(xy11 + z + 1), xr - 1.0, yr - 1.0, zr - 1.0);
    const double xAlpha = smoothstep(xr);
    const double yAlpha = smoothstep(yrOriginal);
    const double zAlpha = smoothstep(zr);
    return lerp3(xAlpha, yAlpha, zAlpha, d000, d100, d010, d110, d001, d101, d011, d111);
}

auto ImprovedNoise::sample_with_derivative(const int x, const int y, const int z, const double xr, const double yr, const double zr, double* derivativeOut) const -> double {
    const int x0 = this->perm(x);
    const int x1 = this->perm(x + 1);
    const int xy00 = this->perm(x0 + y);
    const int xy01 = this->perm(x0 + y + 1);
    const int xy10 = this->perm(x1 + y);
    const int xy11 = this->perm(x1 + y + 1);
    const int p000 = this->perm(xy00 + z);
    const int p100 = this->perm(xy10 + z);
    const int p010 = this->perm(xy01 + z);
    const int p110 = this->perm(xy11 + z);
    const int p001 = this->perm(xy00 + z + 1);
    const int p101 = this->perm(xy10 + z + 1);
    const int p011 = this->perm(xy01 + z + 1);
    const int p111 = this->perm(xy11 + z + 1);
    const int* g000 = SimplexNoise::GRADIENT[p000 & 15];
    const int* g100 = SimplexNoise::GRADIENT[p100 & 15];
    const int* g010 = SimplexNoise::GRADIENT[p010 & 15];
    const int* g110 = SimplexNoise::GRADIENT[p110 & 15];
    const int* g001 = SimplexNoise::GRADIENT[p001 & 15];
    const int* g101 = SimplexNoise::GRADIENT[p101 & 15];
    const int* g011 = SimplexNoise::GRADIENT[p011 & 15];
    const int* g111 = SimplexNoise::GRADIENT[p111 & 15];
    const double d000 = SimplexNoise::dot(g000, xr, yr, zr);
    const double d100 = SimplexNoise::dot(g100, xr - 1.0, yr, zr);
    const double d010 = SimplexNoise::dot(g010, xr, yr - 1.0, zr);
    const double d110 = SimplexNoise::dot(g110, xr - 1.0, yr - 1.0, zr);
    const double d001 = SimplexNoise::dot(g001, xr, yr, zr - 1.0);
    const double d101 = SimplexNoise::dot(g101, xr - 1.0, yr, zr - 1.0);
    const double d011 = SimplexNoise::dot(g011, xr, yr - 1.0, zr - 1.0);
    const double d111 = SimplexNoise::dot(g111, xr - 1.0, yr - 1.0, zr - 1.0);
    const double xAlpha = smoothstep(xr);
    const double yAlpha = smoothstep(yr);
    const double zAlpha = smoothstep(zr);
    const double d1x = lerp3(xAlpha, yAlpha, zAlpha, g000[0], g100[0], g010[0], g110[0], g001[0], g101[0], g011[0], g111[0]);
    const double d1y = lerp3(xAlpha, yAlpha, zAlpha, g000[1], g100[1], g010[1], g110[1], g001[1], g101[1], g011[1], g111[1]);
    const double d1z = lerp3(xAlpha, yAlpha, zAlpha, g000[2], g100[2], g010[2], g110[2], g001[2], g101[2], g011[2], g111[2]);
    const double d2x = lerp2(yAlpha, zAlpha, d100 - d000, d110 - d010, d101 - d001, d111 - d011);
    const double d2y = lerp2(zAlpha, xAlpha, d010 - d000, d011 - d001, d110 - d100, d111 - d101);
    const double d2z = lerp2(xAlpha, yAlpha, d001 - d000, d101 - d100, d011 - d010, d111 - d110);
    const double xSD = smoothstep_derivative(xr);
    const double ySD = smoothstep_derivative(yr);
    const double zSD = smoothstep_derivative(zr);
    const double dX = d1x + xSD * d2x;
    const double dY = d1y + ySD * d2y;
    const double dZ = d1z + zSD * d2z;
    derivativeOut[0] += dX;
    derivativeOut[1] += dY;
    derivativeOut[2] += dZ;
    return lerp3(xAlpha, yAlpha, zAlpha, d000, d100, d010, d110, d001, d101, d011, d111);
}

auto ImprovedNoise::smoothstep(const double x) -> double {
    return x * x * x * (x * (x * 6.0 - 15.0) + 10.0);
}

auto ImprovedNoise::smoothstep_derivative(const double x) -> double {
    return 30.0 * x * x * (x - 1.0) * (x - 1.0);
}

auto ImprovedNoise::lerp(const double alpha1, const double p0, const double p1) -> double {
    return p0 + alpha1 * (p1 - p0);
}

auto ImprovedNoise::lerp2(const double alpha1, const double alpha2, const double x00, const double x10, const double x01, const double x11) -> double {
    return lerp(alpha2, lerp(alpha1, x00, x10), lerp(alpha1, x01, x11));
}

auto ImprovedNoise::lerp3(const double alpha1,
        const double alpha2,
        const double alpha3,
        const double x000,
        const double x100,
        const double x010,
        const double x110,
        const double x001,
        const double x101,
        const double x011,
        const double x111) -> double {
    return lerp(alpha3, lerp2(alpha1, alpha2, x000, x100, x010, x110), lerp2(alpha1, alpha2, x001, x101, x011, x111));
}

File: Fortran/ImprovedNoise.f90
module improved_noise_m
  use, intrinsic :: iso_c_binding
  implicit none
  
  type, bind(c) :: noise_state_t
    real(c_double) :: xo, yo, zo
    integer(c_int8_t) :: p(256)
  end type
  
  integer(c_int), parameter :: GRADIENT(3, 0:15) = reshape(&
    [ 1_c_int, 1_c_int, 0_c_int,  -1_c_int, 1_c_int, 0_c_int, &
      1_c_int,-1_c_int, 0_c_int, -1_c_int,-1_c_int, 0_c_int, &
      1_c_int, 0_c_int, 1_c_int,  -1_c_int, 0_c_int, 1_c_int, &
      1_c_int, 0_c_int,-1_c_int, -1_c_int, 0_c_int,-1_c_int, &
      0_c_int, 1_c_int, 1_c_int,  0_c_int,-1_c_int, 1_c_int, &
      0_c_int, 1_c_int,-1_c_int,  0_c_int,-1_c_int,-1_c_int, &
      1_c_int, 1_c_int, 0_c_int,  0_c_int,-1_c_int, 1_c_int, &
     -1_c_int, 1_c_int, 0_c_int,  0_c_int,-1_c_int,-1_c_int ], &
    [3, 16])
  
contains

  pure function smoothstep(x) result(res)
    real(c_double), intent(in) :: x
    real(c_double) :: res
    res = x*x*x*(x*(6.0_c_double*x - 15.0_c_double) + 10.0_c_double)
  end function

  pure function smoothstep_derivative(x) result(res)
    real(c_double), intent(in) :: x
    real(c_double) :: res, x_m1
    x_m1 = x - 1.0_c_double
    res = 30.0_c_double * x * x * x_m1 * x_m1
  end function

  pure function grad_dot(hash, x, y, z) result(res)
    integer(c_int), intent(in) :: hash
    real(c_double), intent(in) :: x, y, z
    real(c_double) :: res
    integer(c_int) :: h
    
    h = iand(hash, 15)
    res = real(GRADIENT(1, h), c_double) * x + &
          real(GRADIENT(2, h), c_double) * y + &
          real(GRADIENT(3, h), c_double) * z
  end function

  pure function perm(state, x) result(res)
    type(noise_state_t), intent(in) :: state
    integer(c_int), intent(in) :: x
    integer(c_int) :: res
    integer(c_int) :: idx
    
    idx = iand(x, 255) + 1
    res = iand(int(state%p(idx), c_int), 255)
  end function

  pure function lerp(alpha, p0, p1) result(res)
    real(c_double), intent(in) :: alpha, p0, p1
    real(c_double) :: res
    res = p0 + alpha * (p1 - p0)
  end function

  pure function lerp2(a1, a2, x00, x10, x01, x11) result(res)
    real(c_double), intent(in) :: a1, a2, x00, x10, x01, x11
    real(c_double) :: res
    res = lerp(a2, lerp(a1, x00, x10), lerp(a1, x01, x11))
  end function

  pure function lerp3(a1, a2, a3, x000, x100, x010, x110, &
                      x001, x101, x011, x111) result(res)
    real(c_double), intent(in) :: a1, a2, a3
    real(c_double), intent(in) :: x000, x100, x010, x110
    real(c_double), intent(in) :: x001, x101, x011, x111
    real(c_double) :: res
    res = lerp(a3, lerp2(a1, a2, x000, x100, x010, x110), &
                   lerp2(a1, a2, x001, x101, x011, x111))
  end function

  function sample_and_lerperm(state, xf, yf, zf, xr, yr, zr, yrOriginal) &
      result(res)
    type(noise_state_t), intent(in) :: state
    integer(c_int), intent(in) :: xf, yf, zf
    real(c_double), intent(in) :: xr, yr, zr, yrOriginal
    real(c_double) :: res
    
    integer(c_int) :: x0, x1, xy00, xy01, xy10, xy11
    real(c_double) :: d000, d100, d010, d110, d001, d101, d011, d111
    real(c_double) :: xAlpha, yAlpha, zAlpha
    
    x0 = perm(state, xf)
    x1 = perm(state, xf + 1)
    xy00 = perm(state, x0 + yf)
    xy01 = perm(state, x0 + yf + 1)
    xy10 = perm(state, x1 + yf)
    xy11 = perm(state, x1 + yf + 1)
    
    d000 = grad_dot(perm(state, xy00 + zf), xr, yr, zr)
    d100 = grad_dot(perm(state, xy10 + zf), xr - 1.0_c_double, yr, zr)
    d010 = grad_dot(perm(state, xy01 + zf), xr, yr - 1.0_c_double, zr)
    d110 = grad_dot(perm(state, xy11 + zf), xr - 1.0_c_double, yr - 1.0_c_double, zr)
    d001 = grad_dot(perm(state, xy00 + zf + 1), xr, yr, zr - 1.0_c_double)
    d101 = grad_dot(perm(state, xy10 + zf + 1), xr - 1.0_c_double, yr, zr - 1.0_c_double)
    d011 = grad_dot(perm(state, xy01 + zf + 1), xr, yr - 1.0_c_double, zr - 1.0_c_double)
    d111 = grad_dot(perm(state, xy11 + zf + 1), xr - 1.0_c_double, yr - 1.0_c_double, zr - 1.0_c_double)
    
    xAlpha = smoothstep(xr)
    yAlpha = smoothstep(yrOriginal)
    zAlpha = smoothstep(zr)
    
    res = lerp3(xAlpha, yAlpha, zAlpha, d000, d100, d010, d110, &
                d001, d101, d011, d111)
  end function

  function sample_with_derivative(state, xf, yf, zf, xr, yr, zr, deriv) &
      result(res)
    type(noise_state_t), intent(in) :: state
    integer(c_int), intent(in) :: xf, yf, zf
    real(c_double), intent(in) :: xr, yr, zr
    real(c_double), intent(inout) :: deriv(3)
    real(c_double) :: res
    
    integer(c_int) :: x0, x1, xy00, xy01, xy10, xy11
    integer(c_int) :: p000, p100, p010, p110, p001, p101, p011, p111
    integer(c_int) :: g000(3), g100(3), g010(3), g110(3)
    integer(c_int) :: g001(3), g101(3), g011(3), g111(3)
    real(c_double) :: d000, d100, d010, d110, d001, d101, d011, d111
    real(c_double) :: xAlpha, yAlpha, zAlpha
    real(c_double) :: d1x, d1y, d1z, d2x, d2y, d2z
    real(c_double) :: xSD, ySD, zSD, dX, dY, dZ
    
    x0 = perm(state, xf)
    x1 = perm(state, xf + 1)
    xy00 = perm(state, x0 + yf)
    xy01 = perm(state, x0 + yf + 1)
    xy10 = perm(state, x1 + yf)
    xy11 = perm(state, x1 + yf + 1)
    
    p000 = perm(state, xy00 + zf); g000 = GRADIENT(:, iand(p000, 15))
    p100 = perm(state, xy10 + zf); g100 = GRADIENT(:, iand(p100, 15))
    p010 = perm(state, xy01 + zf); g010 = GRADIENT(:, iand(p010, 15))
    p110 = perm(state, xy11 + zf); g110 = GRADIENT(:, iand(p110, 15))
    p001 = perm(state, xy00 + zf + 1); g001 = GRADIENT(:, iand(p001, 15))
    p101 = perm(state, xy10 + zf + 1); g101 = GRADIENT(:, iand(p101, 15))
    p011 = perm(state, xy01 + zf + 1); g011 = GRADIENT(:, iand(p011, 15))
    p111 = perm(state, xy11 + zf + 1); g111 = GRADIENT(:, iand(p111, 15))
    
    d000 = real(g000(1), c_double)*xr + real(g000(2), c_double)*yr + real(g000(3), c_double)*zr
    d100 = real(g100(1), c_double)*(xr-1.0_c_double) + real(g100(2), c_double)*yr + real(g100(3), c_double)*zr
    d010 = real(g010(1), c_double)*xr + real(g010(2), c_double)*(yr-1.0_c_double) + real(g010(3), c_double)*zr
    d110 = real(g110(1), c_double)*(xr-1.0_c_double) + real(g110(2), c_double)*(yr-1.0_c_double) + real(g110(3), c_double)*zr
    d001 = real(g001(1), c_double)*xr + real(g001(2), c_double)*yr + real(g001(3), c_double)*(zr-1.0_c_double)
    d101 = real(g101(1), c_double)*(xr-1.0_c_double) + real(g101(2), c_double)*yr + real(g101(3), c_double)*(zr-1.0_c_double)
    d011 = real(g011(1), c_double)*xr + real(g011(2), c_double)*(yr-1.0_c_double) + real(g011(3), c_double)*(zr-1.0_c_double)
    d111 = real(g111(1), c_double)*(xr-1.0_c_double) + real(g111(2), c_double)*(yr-1.0_c_double) + real(g111(3), c_double)*(zr-1.0_c_double)
    
    xAlpha = smoothstep(xr)
    yAlpha = smoothstep(yr)
    zAlpha = smoothstep(zr)
    
    d1x = lerp3(xAlpha, yAlpha, zAlpha, &
                real(g000(1), c_double), real(g100(1), c_double), &
                real(g010(1), c_double), real(g110(1), c_double), &
                real(g001(1), c_double), real(g101(1), c_double), &
                real(g011(1), c_double), real(g111(1), c_double))
    d1y = lerp3(xAlpha, yAlpha, zAlpha, &
                real(g000(2), c_double), real(g100(2), c_double), &
                real(g010(2), c_double), real(g110(2), c_double), &
                real(g001(2), c_double), real(g101(2), c_double), &
                real(g011(2), c_double), real(g111(2), c_double))
    d1z = lerp3(xAlpha, yAlpha, zAlpha, &
                real(g000(3), c_double), real(g100(3), c_double), &
                real(g010(3), c_double), real(g110(3), c_double), &
                real(g001(3), c_double), real(g101(3), c_double), &
                real(g011(3), c_double), real(g111(3), c_double))
    
    d2x = lerp2(yAlpha, zAlpha, d100 - d000, d110 - d010, d101 - d001, d111 - d011)
    d2y = lerp2(zAlpha, xAlpha, d010 - d000, d011 - d001, d110 - d100, d111 - d101)
    d2z = lerp2(xAlpha, yAlpha, d001 - d000, d101 - d100, d011 - d010, d111 - d110)
    
    xSD = smoothstep_derivative(xr)
    ySD = smoothstep_derivative(yr)
    zSD = smoothstep_derivative(zr)
    
    dX = d1x + xSD * d2x
    dY = d1y + ySD * d2y
    dZ = d1z + zSD * d2z
    
    deriv(1) = deriv(1) + dX
    deriv(2) = deriv(2) + dY
    deriv(3) = deriv(3) + dZ
    
    res = lerp3(xAlpha, yAlpha, zAlpha, d000, d100, d010, d110, &
                d001, d101, d011, d111)
  end function

  
  !DEC$ ATTRIBUTES DLLEXPORT :: ImprovedNoise_noise_3
  function ImprovedNoise_noise_3(state_ptr, x, y, z) bind(c, name="ImprovedNoise_noise_3")
    type(c_ptr), value :: state_ptr
    real(c_double), value :: x, y, z
    real(c_double) :: ImprovedNoise_noise_3
    type(noise_state_t), pointer :: state
    
    call c_f_pointer(state_ptr, state)
    ImprovedNoise_noise_3 = ImprovedNoise_noise_5(state_ptr, x, y, z, 0.0_c_double, 0.0_c_double)
  end function

  !DEC$ ATTRIBUTES DLLEXPORT :: ImprovedNoise_noise_5
  function ImprovedNoise_noise_5(state_ptr, x, y, z, yScale, yFudge) bind(c, name="ImprovedNoise_noise_5")
    type(c_ptr), value :: state_ptr
    real(c_double), value :: x, y, z, yScale, yFudge
    real(c_double) :: ImprovedNoise_noise_5
    
    type(noise_state_t), pointer :: state
    real(c_double) :: x_adj, y_adj, z_adj, xr, yr, zr, yr_fudge, fudge_limit
    integer(c_int) :: xf, yf, zf
    
    call c_f_pointer(state_ptr, state)
    
    x_adj = x + state%xo
    y_adj = y + state%yo
    z_adj = z + state%zo
    
    xf = floor(x_adj)
    yf = floor(y_adj)
    zf = floor(z_adj)
    xr = x_adj - xf
    yr = y_adj - yf
    zr = z_adj - zf
    
    if (yScale /= 0.0_c_double) then
      if (yFudge >= 0.0_c_double .and. yFudge < yr) then
        fudge_limit = yFudge
      else
        fudge_limit = yr
      end if
      yr_fudge = floor(fudge_limit / yScale + 1.0e-7_c_double) * yScale
    else
      yr_fudge = 0.0_c_double
    end if
    
    ImprovedNoise_noise_5 = sample_and_lerperm(state, xf, yf, zf, xr, yr - yr_fudge, zr, yr)
  end function

  !DEC$ ATTRIBUTES DLLEXPORT :: ImprovedNoise_noise_with_derivative
  function ImprovedNoise_noise_with_derivative(state_ptr, x, y, z, deriv_ptr) bind(c, name="ImprovedNoise_noise_with_derivative")
    type(c_ptr), value :: state_ptr
    real(c_double), value :: x, y, z
    type(c_ptr), value :: deriv_ptr
    real(c_double) :: ImprovedNoise_noise_with_derivative
    
    type(noise_state_t), pointer :: state
    real(c_double), pointer :: deriv(:)
    real(c_double) :: x_adj, y_adj, z_adj, xr, yr, zr
    integer(c_int) :: xf, yf, zf
    
    call c_f_pointer(state_ptr, state)
    call c_f_pointer(deriv_ptr, deriv, [3])
    
    x_adj = x + state%xo
    y_adj = y + state%yo
    z_adj = z + state%zo
    
    xf = floor(x_adj)
    yf = floor(y_adj)
    zf = floor(z_adj)
    xr = x_adj - xf
    yr = y_adj - yf
    zr = z_adj - zf
    
    ImprovedNoise_noise_with_derivative = sample_with_derivative(state, xf, yf, zf, xr, yr, zr, deriv)
  end function

end module